FAQ node api doc for kms systems (#224)
hellohaptik/docs_haptik: docs/bot-builder/integrations/additional-apis.md

}
```

## FAQ Node Creation APIs for KMS

> We have many clients who uses FAQ bots, So, everytime when we want to change Q&A, a new csv with node name, question and answer in certain format(given below) needs to be uploaded.

Reviewer: Please elaborate it a little more like on the lines of -
We have many clients who need FAQ bots. Every time when we want to add/replace/change the content of those FAQs, a new CSV with Node name, Question and Answer in a defined format needs to be uploaded. [Format given below]

> We have many clients who uses FAQ bots, So, everytime when we want to change Q&A, a new csv with node name, question and answer in certain format(given below) needs to be uploaded.

Author: We have many clients who uses FAQ bots, So, everytime when we want to change Q&A, a new csv with node name, question and answer in certain format(given below) needs to be uploaded.

> We have many clients who uses FAQ bots, So, everytime when we want to change Q&A, a new csv with node name, question and answer in certain format(given below) needs to be uploaded.

Author: ```suggestion
We have many clients who need FAQ bots. Every time when we want to change the content of these FAQs (Q&As), a new CSV with Node name, Question and Answer in a defined format needs to be uploaded. (Format given below)
```

> Although, frequency of this activity varies from business to business but it's a repeatable to smoothen this tedious process we've developed an API which can be integrated with the client KMS(knowledge Management Systems) via any intermediate layer and everytime when data gets changed in KMS, KMS should get call this API to keep data in sync with the haptik systems.

Reviewer: We have developed an API which can be integrated with the client KMS(Knowledge Management Systems) via any intermediate layer and every time when data gets changed in KMS, it should call this API to keep data in sync with the haptik systems.

> Although, frequency of this activity varies from business to business but it's a repeatable to smoothen this tedious process we've developed an API which can be integrated with the client KMS(knowledge Management Systems) via any intermediate layer and everytime when data gets changed in KMS, KMS should get call this API to keep data in sync with the haptik systems.

Author: ```suggestion
We have developed an API that can be integrated with the client KMS(Knowledge Management Systems) via any intermediate layer and every time when data gets changed in KMS, it should call this API to keep data in sync with the Haptik systems.
```

This API allows you to create/update/delete faq nodes via a `POST` request to Haptik Platform. 

> Note: The `base-url` will be provided by Haptik at the time of integration.

FAQ Node creation process API consists of two parts:

> 1. **Submit the CSV**: To handle the large csv files, this API takes the request data and after validation submit it to celery worker and returns the `task id`, which executes the task asynchronously.

Reviewer: ```suggestion
1. **Submit the CSV**: To handle the large csv files, this API takes the request data and after backend validation submits it to celery worker and returns the `task id`, which executes the task asynchronously.
```

2. **Polling API** to fetch the status of submitted `task id`

Example URL: `https://<base-url>/mogambo_api/nodes/faq/create/`

![CSV Format](assets/faq_node_csv_format.png)

```
1. node_name: Name of the node, node name should be unique

> 2. question(user says): user says to detect this faq node, multiple user says for single node can be passed as pipe(|)  separated values(ref the sample row in image above)

Reviewer: ```suggestion
2. Question(User Says): Independent Responses under User Says to detect this FAQ node. Multiple user says for single node can be passed as pipe(|) separated values(refer the sample row in image above)
```

3. answer(bot says): A node can have single or multiple bot says, to pass multiple bot says use `<m>` tag as delimiter
```

#### Headers

```
Authorization: Bearer <TOKEN>
client-id: <CLIENT_ID>
Content-Type: multipart/form-data
```

- Authorization - The Authorization header of each HTTP request should be “Bearer” followed by your token which will be provided by Haptik
- client-id - The client id for your account which will be provided by Haptik
- Content-Type - multipart/form-data

#### Request Body

```form-data
"domain_name": "<Name of the Domain/Bot>",

> "csv_file": "csv file",

Reviewer: ```suggestion
"csv_file": filename.csv,
```

Reviewer: ```suggestion
"csv_file": "filename.csv",
```

```

#### Response

A successful request to the API will return a `200` status code with a JSON response object.

```json
{
    "success": true,
    "body": {
        "task_id": "fe139ce5-f4c3-4383-bacb-64a546cb0b6c"
    },
    "error": "",
    "meta": {}
}
```

- success flag Indicates if the API was a success or failure
- task_id: Job id for the submitted task

#### Error Response

If there is any error in the Headers or the Request body, then the Error message will be returned in a JSON as shown below:

```json
{
  "error_message": "invalid doman"
}
```

| Error Code | Error Message|
|----------|:-------------:|
| 401 | Unauthorized Access or invalid token|
| 400 | Invalid or bad request body   |
| 403 | Access Forbidden |

#### Sample Curl Command to submit csv file and get job id
```
curl -X POST \
  https://<base-url>/mogambo_api/nodes/faq/create/ \
  -H 'authorization: Bearer <token>' \
  -H 'client-id: <client-id>' \
  -H 'content-type: multipart/form-data; 
  -F 'domain_name=<domain_name>' \

>   -F csv_file=@csv_file.csv

Reviewer: ```suggestion
  -F csv_file=@filename.csv
```

```

> > Note: This API follows the upsert behavior viz everytime when new csv gets uploaded it make the existing faq nodes for that domain inactive and create new nodes for the rows mentioned in the CSV

Reviewer: ```suggestion
> Note: This API follows the upsert [UPDATE and INSERT] behaviour i.e. every time when new CSV gets uploaded it makes the existing FAQ nodes for that domain inactive, creates new nodes for the rows mentioned in the CSV and updates Nodes with same names with the new User Says and Bot Says from the CSV. 
```

> ## Polling API to fetch submitted task Status

Reviewer: Please put links at relevant places
